File: services/agent-gateway/src/request-workflow-log-store.ts
import type { RequestContext, WorkflowStep } from '@openAwork/logger';
import { sqliteAll, sqliteRun } from './db.js';
import { isSqliteMalformedError } from './sqlite-error-utils.js';

interface RequestWorkflowLogRow {
  id: number;
  request_id: string;
  user_id: string | null;
  session_id: string | null;
  method: string;
  path: string;
  status_code: number;
  ip: string | null;
  user_agent: string | null;
  workflow_json: string;
  created_at: string;
}

let requestWorkflowLogStoreDisabled = false;

function stripPrivateWorkflowFields(step: WorkflowStep): Omit<WorkflowStep, '_startedAt'> {
  return {
    name: step.name,
    status: step.status,
    ...(step.message !== undefined ? { message: step.message } : {}),
    ...(step.durationMs !== undefined ? { durationMs: step.durationMs } : {}),
    ...(step.fields ? { fields: step.fields } : {}),
    ...(step.children ? { children: step.children.map(stripPrivateWorkflowFields) } : {}),
  };
}

function detectSessionId(path: string): string | null {
  const match = path.match(/\/sessions\/([^/?]+)/u);
  return match?.[1] ?? null;
}

export function persistRequestWorkflowLog(input: {
  context: RequestContext;
  steps: WorkflowStep[];
  statusCode: number;
  userId?: string | null;
}): void {
  if (requestWorkflowLogStoreDisabled) {
    return;
  }

  try {
    sqliteRun(
      `INSERT INTO request_workflow_logs
       (request_id, user_id, session_id, method, path, status_code, ip, user_agent, workflow_json, created_at)
       VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, datetime('now'))`,
      [
        input.context.requestId,
        input.userId ?? null,
        detectSessionId(input.context.path),
        input.context.method,
        input.context.path,
        input.statusCode,
        input.context.ip ?? null,
        input.context.userAgent ?? null,
        JSON.stringify(input.steps.map(stripPrivateWorkflowFields)),
      ],
    );
  } catch (error) {
    if (isSqliteMalformedError(error)) {
      requestWorkflowLogStoreDisabled = true;
      return;
    }

    throw error;
  }
}

export function listRequestWorkflowLogs(userId: string, limit = 100): RequestWorkflowLogRow[] {
  if (requestWorkflowLogStoreDisabled) {
    return [];
  }

  try {
    return sqliteAll<RequestWorkflowLogRow>(
      `SELECT id, request_id, user_id, session_id, method, path, status_code, ip, user_agent, workflow_json, created_at
       FROM request_workflow_logs
       WHERE user_id = ?
       ORDER BY created_at DESC
       LIMIT ?`,
      [userId, limit],
    );
  } catch (error) {
    if (isSqliteMalformedError(error)) {
      requestWorkflowLogStoreDisabled = true;
      return [];
    }

    throw error;
  }
}

export function resetRequestWorkflowLogStoreStateForTests(): void {
  requestWorkflowLogStoreDisabled = false;
}

File: services/agent-gateway-dotnet/src/OpenAWork.Gateway.Persistence.EFCore/Stores/RequestWorkflowLogStore.cs
using System.Text.Json;
using Microsoft.EntityFrameworkCore;
using OpenAWork.Gateway.Application.Abstractions.Observability;
using OpenAWork.Gateway.Contracts.Observability;
using OpenAWork.Gateway.Persistence.EFCore.Entities;

namespace OpenAWork.Gateway.Persistence.EFCore.Stores;

public sealed class RequestWorkflowLogStore(GatewayDbContext dbContext) : IRequestWorkflowLogStore
{
    private static readonly JsonSerializerOptions JsonOptions = new()
    {
        PropertyNamingPolicy = JsonNamingPolicy.CamelCase,
        DefaultIgnoreCondition = System.Text.Json.Serialization.JsonIgnoreCondition.WhenWritingNull,
    };

    public async Task PersistAsync(
        RequestContext context,
        IReadOnlyCollection<WorkflowStep> steps,
        int statusCode,
        string? userId,
        CancellationToken cancellationToken)
    {
        var record = new RequestWorkflowLogRecord
        {
            RequestId = context.RequestId,
            UserId = userId,
            SessionId = DetectSessionId(context.Path),
            Method = context.Method,
            Path = context.Path,
            StatusCode = statusCode,
            Ip = context.Ip,
            UserAgent = context.UserAgent,
            WorkflowJson = JsonSerializer.Serialize(steps.Select(CloneStep), JsonOptions),
            CreatedAtUtc = DateTimeOffset.UtcNow,
        };

        dbContext.RequestWorkflowLogs.Add(record);
        await dbContext.SaveChangesAsync(cancellationToken);
    }

    public async Task<IReadOnlyList<RequestWorkflowLogEntry>> ListByUserAsync(string userId, int limit, CancellationToken cancellationToken)
    {
        return await dbContext.RequestWorkflowLogs
            .AsNoTracking()
            .Where((record) => record.UserId == userId)
            .OrderByDescending((record) => record.Id)
            .Take(limit)
            .Select((record) => new RequestWorkflowLogEntry(
                record.Id,
                record.RequestId,
                record.UserId,
                record.SessionId,
                record.Method,
                record.Path,
                record.StatusCode,
                record.WorkflowJson,
                record.CreatedAtUtc))
            .ToListAsync(cancellationToken);
    }

    public Task<int> DeleteErrorLogsByUserAsync(string userId, CancellationToken cancellationToken)
    {
        return dbContext.RequestWorkflowLogs
            .Where((record) => record.UserId == userId && record.StatusCode >= 400)
            .ExecuteDeleteAsync(cancellationToken);
    }

    private static string? DetectSessionId(string path)
    {
        var marker = "/sessions/";
        var startIndex = path.IndexOf(marker, StringComparison.OrdinalIgnoreCase);
        if (startIndex < 0)
        {
            return null;
        }

        var remainder = path[(startIndex + marker.Length)..];
        var endIndex = remainder.IndexOf('/');
        return endIndex >= 0 ? remainder[..endIndex] : remainder;
    }

    private static WorkflowStep CloneStep(WorkflowStep step)
    {
        return new WorkflowStep
        {
            Name = step.Name,
            Status = step.Status,
            Message = step.Message,
            DurationMs = step.DurationMs,
            Fields = step.Fields is null ? null : new Dictionary<string, string>(step.Fields),
            StartedAtUtc = null,
            Children = step.Children.Select(CloneStep).ToList(),
        };
    }
}
